llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -57,197 +57,121 @@ static cl::opt<bool>
+  const TargetRegisterInfo &TRI;
+
+public:
+  PhysRegCopyTracker(const MachineRegisterInfo &MRI)
+      : TRI(*MRI.getTargetRegisterInfo()) {
+    assert(MRI.isSSA() && "PhysRegCopyTracker can only track SSA copies");
   }
-  return Changed;
-}
 
-/// Check if the virtual register being copied to the phys reg was copied from
-/// from the same phys reg.
-bool AIEPostSelectOptimize::isCopyFromPhysReg(const MachineInstr &MI,
-                                              const Register PhysReg) {
-  if (!MI.isCopy())
-    return false;
-  Register SrcReg = MI.getOperand(1).getReg();
-  if (SrcReg.isPhysical())
-    return SrcReg == PhysReg;
-  MachineInstr *VRegDef = MRI->getVRegDef(SrcReg);
-  if (!VRegDef)
-    return false;
-  return isCopyFromPhysReg(*VRegDef, PhysReg);
-}
+  /// Invalidate the copies of \p Reg.
+  void invalidateCopies(MCRegister Reg) { Copies.erase(Reg); }
 
-/// Stores the first identity copy to act as anchor for all the other copies and
-/// remove later if applicable.
-void AIEPostSelectOptimize::storeFirstIdentityCopy(MachineInstr &MI) {
-  assert(MI.getNumOperands() == 2 && "Expected only 2 operands in the MI");
-
-  if (!MI.getOperand(1).isReg())
-    return;
-
-  Register DstPhysReg = MI.getOperand(0).getReg();
-  Register SrcReg = MI.getOperand(1).getReg();
-  // We currently have $PhysReg = COPY SrcReg. Check if the SrcReg is virtual
-  // and find the def of the SrcReg to check for %virtreg = COPY $PhysReg.
-  if (ErasableVirtRegToPhysRegCopies.count(SrcReg) ||
-      !isCopyFromPhysReg(MI, DstPhysReg)) {
-    return;
+  /// Invalidate copies for any reg defined by \p MI
+  void invalidateDefCopies(const MachineInstr &MI) {
+    for (const MachineOperand &MO : MI.all_defs()) {
+      if (MO.getReg().isPhysical()) {
+        invalidateCopies(MO.getReg());
+      }
+    }
   }
 
-  // If the copy is found and copy propagation will lead to identity copy,
-  // then just store the current $PhysReg = COPY %virtreg MI to be removed
-  // later.
-  ErasableVirtRegToPhysRegCopies[SrcReg] = &MI;
-}
+  /// Track that \p VRegCopy is a copy of \p PhysReg
+  void trackCopy(MCRegister PhysReg, Register VRegCopy) {
+    assert(VRegCopy.isVirtual());
+    assert(range_size(TRI.regunits(PhysReg)) == 1 && "Phys reg has aliases.");
+    Copies[PhysReg].insert(VRegCopy);
+  }
 
-/// If there is an intervening clobber but copy propagation suggests it
-/// comes from the same PhysReg, it can be removed.
-bool AIEPostSelectOptimize::eraseSimilarClobber(MachineInstr &MI,
-                                                MachineInstr &LastVisitedDef) {
-  if (!isCopyFromPhysReg(LastVisitedDef, MI.getOperand(0).getReg()) ||
-      !isCopyFromPhysReg(MI, LastVisitedDef.getOperand(0).getReg()))
+  /// Return true if \p VReg is a copy of \p PhysReg
+  bool isCopy(MCRegister PhysReg, Register VReg) const {
+    if (auto It = Copies.find(PhysReg); It != Copies.end())
+      return It->second.contains(VReg);
```

Comment:
nit: These two interfaces record and check VReg = COPY(PhysReg); I would swap the parameters